### PR TITLE
Introduction to Dafny: Insertion Sort 

### PR DESCRIPTION
**Pull request by jtristan**
* New build system using Python and Madoko
* New blog post on insertion sort

### CODE REVIEW

File: assets/mdk/insertion-sort.mdk [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+
+The key characteristic that makes Dafny quite different from most
+other programming languages is that it is designed with formal
+verification in mind. In Dafny, you can annotate your code to
+mathematically formalize its specification, and verify that your code
+is correct with respect to its specification. Dafny is not just
+designed to support object-oriented programming and functional
+programming, but also verified programming.
+
+We're going to introduce verified programming in Dafny will give you a
```
Comment (atomb) [inline assets/mdk/insertion-sort.mdk:34]: This sentence looks like a victim of incomplete editing.

File: assets/mdk/insertion-sort.mdk [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+Dafny is a full-fledged programming language and you may even want to
+use it if you have no interest in verification whatsoever. In fact,
+writing unverified Dafny program is a good way to get familiarity with
+the language.  Following is one possible implementation of
+insertion sort in Dafny.
+
+<!-- inline-dafny Unverified/M1 -->
+
+Insertion sort is implemented as a method named `InsertionSort` (which
+needs not belong to a class) that takes as a input parameter an array
```
Comment (atomb) [inline assets/mdk/insertion-sort.mdk:50]:
```suggestion
need not belong to a class) that takes as an input parameter an array
```

File: assets/mdk/insertion-sort.mdk [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+formula. If we can prove that this mathematical formula is true, then
+the code satisfies the specification. Otherwise it may or may
+not. Proving such mathematical formula would be a lot of work, so
+Dafny makes use of an *automated theorem prover* (ATP) to prove it on
+your behalf.
+
+Alas, ATPs have fundamental limitations and while they excel at
+proving properties of linear code, they need some help with
+loops. Fundamentally, they need to know of properties that are true
+when the loop starts its execution ; remain true after execution of
```
Comment (atomb) [inline assets/mdk/insertion-sort.mdk:131]:
```suggestion
when the loop starts its execution, remain true after execution of
```

File: assets/mdk/insertion-sort.mdk [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+the loop ; imply what needs to be true after execution of the
```
Comment (atomb) [inline assets/mdk/insertion-sort.mdk:132]:
```suggestion
the loop, and imply what needs to be true after execution of the
```

File: assets/mdk/insertion-sort.mdk [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+properties that characterize the functional behavior of a sorting
+algorithm, one of which is that once the computation is done, the
+values in the array should be ordered. Before we can do the actual
+verification, we need to specify what we expect of insertion
+sort. First, we need to formalize what it means for an array to be
+ordered.
+
+In English, we might say something like: *Definition: an array is
+ordered if for any non-negative integer value i ranging from 0 to the
+length of the array, excluded, the value at index i is less than the
```
Comment (atomb) [inline assets/mdk/insertion-sort.mdk:162]:
```suggestion
length of the array, exclusive, the value at index i is less than the
```

File: assets/mdk/insertion-sort.mdk [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+The second clause, `requires left <= right <= a.Length`, is a
+precondition that restricts the definition to values of `left`,
+`right`, and `a.Length` to ones satisfying that precondition. Such
+preconditions are a fundamental feature of Dafny that allow for the
+definition of partial functions. It may not be obvious at first sight,
+but this is highly valuable to writing efficient code, since you can
+define partial functions and ensure statically that they will not be
+used outside of their intended domain of definition, removing any need
+for checking arguments at runtime and making your code throw an
+exception (recall: Dafny does not have exceptions!), or worse, return a
```
Comment (atomb) [inline assets/mdk/insertion-sort.mdk:208]:
```suggestion
exception (recall: Dafny does not have exceptions!), or worse, returning a
```

**Review by atomb — CHANGES_REQUESTED** [review]
Very nice! I noticed a few typos and places where wording could be improved, but otherwise think it looks good as-is.

**Review by MikaelMayer — APPROVED** [review]
Excellent blog post thanks !

**Review by atomb — APPROVED** [review]

**Pull request merged**